src/transformers/generation_tf_utils.py

```diff
@@ -677,21 +685,35 @@ def generate(
             # get encoder and store encoder outputs
             encoder = self.get_encoder()
 
-            encoder_outputs = encoder(
-                input_ids,
-                attention_mask=attention_mask,
-                output_attentions=output_attentions,
-                output_hidden_states=output_hidden_states,
-                return_dict=return_dict_in_generate,
-            )
+            encoder_kwargs = {
+                "attention_mask": attention_mask,
+                "output_attentions": output_attentions,
+                "output_hidden_states": output_hidden_states,
+                "return_dict": return_dict_in_generate,
+            }
+
+            # vision models don't use `attention_mask`.
+            # PT's generate() also requires a similar change, at the following line
+            #   "if model_kwargs.get("attention_mask", None) is None:"
+            # TODO: change PT's generate() and remove `attention_mask` from `ViTModel`
+            signature = dict(inspect.signature(encoder.call).parameters)
+            if "attention_mask" not in signature:
+                encoder_kwargs.pop("attention_mask")
+
+            encoder_outputs = encoder(input_ids, **encoder_kwargs)
             if return_dict_in_generate:
                 if output_attentions:
                     model_kwargs["encoder_attentions"] = encoder_outputs.attentions
                 if output_hidden_states:
                     model_kwargs["encoder_hidden_states"] = encoder_outputs.hidden_states
 
+        # Add an extra condition `len(shape_list(input_ids)) == 2`, so this block will treat only text inputs.
+        # (vision inputs might occur when the model is an encoder-decoder model)
+        # In PT generate(), the `input_ids` is reset for decoder at the following line, which is before the expansion:
+        #   "# set input_ids as decoder_input_ids"
+        # TODO: it would be better to have the same approach as in PT's generate()
```

Comment:
Reformulate the comment to just explain why there is a `len(shape_list(input_ids)) == 2` here.